default/python/universe_generation/monsters.py

```diff
@@ -1,8 +1,110 @@
+        if fo.create_monster(design, monster_fleet) == fo.invalid_object():
+            errstr = "Python generate_monsters: unable to create monster %s" % design
+            util.report_error(errstr)
+            raise RuntimeError(errstr)
+
+    print "Spawn", fleet_plan.name(), "at", fo.get_name(system)
+
+monsters_that_alter_starlanes = {"SM_EXP_OUTPOST"}
+
+
+def place_experimentors(systems):
+    """
+    Place the expermentor outpost system.  Check that removing all of the
+    starlanes entering the system does not break the rest of the galaxy
+    into more than one connected piece.
+    """
+
+    random.shuffle(systems)
+
+    # Find the Experimentor output fleet plan
+    plans = [plan for plan in fo.load_monster_fleet_plan_list()
+             if (plan.name() in monsters_that_alter_starlanes
+                 and plan.spawn_rate() > 0.0
+                 and plan.spawn_limit() > 0
+                 and plan.ship_designs())]
+    if not plans:
+        return
+
+    plan = plans[0]
+    spawn_limit = plan.spawn_limit()
+    placed = []
+    removed_lanes = set()
+
+    starlanes = [(s1, s2) for s1 in systems for s2 in fo.sys_get_starlanes(s1) if s1 < s2]
+
+    for system in systems:
+        # Quit when no more to spawn
+        if len(placed) >= spawn_limit:
+            return
+
+        # Continue if system is not suitable.
+        if not plan.location(system):
+            continue
+
+        # Compute the disjoint set of the galaxy without the starlanes
+        # from the proposed system.  Don't place an experimentor outpost
+        # if there will be more connected regions than the number of
+        # placed experimentor outposts plus one.
+        local_lanes = {(min(system, s), max(system, s)) for s in fo.sys_get_starlanes(system)}
+
+        dsets = DisjointSets()
+        for s3 in systems:
+            dsets.add(s3)
+
+        for lane in starlanes:
+            if lane in removed_lanes or lane in local_lanes:
+                continue
+            dsets.link(lane[0], lane[1])
+
+        num_contiguous_regions = len(dsets.complete_sets())
+        expected_num_contiguous_regions = (len(placed) + 2)
+
+        if num_contiguous_regions > expected_num_contiguous_regions:
+            continue
+
+        if num_contiguous_regions < expected_num_contiguous_regions:
+            print >> sys.stderr, ("Number of contiguos regions %d is below the expected number "
+                                  "of contiguous regions %d when placing %d monster %s that can "
+                                  "break starlanes."
+                                  % (num_contiguous_regions, expected_num_contiguous_regions,
+                                     len(placed) + 1, plan.name()))
+            continue
+
+        # create monster fleet
+        try:
+            populate_monster_fleet(plan, system)
+            placed.append(system)
```

Comment:
Again, correct me if I'm wrong, but won't break that code if someone just raises the spawn limit for the experimentor monster? It might be highly unlikely, but `placed` is a list, and in case two or more experimentor monsters are placed within the same system, `len(placed)` will count the same system multiple times, and your calculation of `expected_num_contiguous_regions` will go wrong, won't it?